File: backend/src/persistence/entities/semgrep-rule-set.entity.ts
import { Column, Entity, Index, PrimaryGeneratedColumn } from 'typeorm';
import { bigIntColumn, boolColumn, intColumn, jsonColumn, stringColumn, textColumn, timestampColumn } from '../columns';

/**
 * A Semgrep rule set uploaded by an operator.
 *
 * Zanshin bundles one rule; the public sets are not redistributable, so coverage arrives
 * from outside ([decision 0006](../../../../docs/architecture/decisions/0006-semgrep-rules-written-here.md)).
 * `ZANSHIN_SEMGREP_RULES_DIR` is one route and it is per-executor: every remote agent needs
 * the directory on its own filesystem, and the control plane cannot verify that it did.
 * Storing the rules here and letting every executor fetch them makes them identical by
 * construction.
 *
 * **The files live in one JSON column rather than one row each.** A rule set is written
 * once, read whole, and never queried by file: a row per file would buy a join and a
 * consistency problem — a fetch landing between two inserts would ship half a rule set,
 * which is exactly the silent partial coverage this table exists to prevent.
 *
 * `contentHash` is the identity an executor caches on, so an unchanged set is fetched once
 * rather than on every claim.
 */
@Entity('t_semgrep_rule_set')
@Index('uq_semgrep_rule_set_active', ['isActive'], { unique: true })
export class SemgrepRuleSet {
    @PrimaryGeneratedColumn()
    id!: number;

    /** The operator's name for this set. Display only. */
    @Column(stringColumn(255))
    name!: string;

    /**
     * `{ path, originalName, content }[]`, under the paths Zanshin generated.
     *
     * The uploaded names are kept alongside so a rule can be traced back to the file it came
     * from; they are never used as paths — see `domain/rules/rule-set.ts`.
     */
    @Column(jsonColumn())
    files!: unknown;

    /** SHA-256 over the generated paths and contents. What an executor caches on. */
    @Column({ ...stringColumn(64), name: 'content_hash' })
    contentHash!: string;

    /** How many rule ids the advisory extraction found. Display only. */
    @Column({ ...intColumn(), name: 'rule_count' })
    ruleCount!: number;

    @Column({ ...intColumn(), name: 'file_count' })
    fileCount!: number;

    @Column({ ...bigIntColumn(), name: 'size_bytes' })
    sizeBytes!: string;

    /**
     * At most one set is active, and it is the one that ships with every task.
     *
     * **Nullable rather than a plain boolean**, and that is the portable way to say "at most
     * one true" across these four engines: a unique index over `is_active` counts `NULL`s as
     * distinct on all of them, so inactive rows carry `NULL` and the active one carries
     * `true`. A `false` there would make the second deactivated set collide with the first.
     */
    @Column({ ...boolColumn({ nullable: true }), name: 'is_active' })
    isActive!: boolean | null;

    @Column({ ...stringColumn(255, { nullable: true }), name: 'uploaded_by' })
    uploadedBy!: string | null;

    @Column({ ...timestampColumn(), name: 'uploaded_at' })
    uploadedAt!: Date;

    /**
     * What the operator was told they were about to lose, kept as a record of the decision.
     *
     * Activating a set whose rule ids differ resolves the open issues of the rules that
     * disappear — with their triage decisions. Recording the warning that was shown makes
     * "why did four hundred issues close that afternoon" answerable six months later.
     */
    @Column({ ...textColumn({ nullable: true }), name: 'activation_note' })
    activationNote!: string | null;
}

File: vectispire-java/vectispire-core/src/main/java/com/asmolabs/vectispire/core/persistence/SemgrepRuleSetEntity.java
package com.asmolabs.vectispire.core.persistence;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

/**
 * An uploaded rule set.
 *
 * <p>{@code isActive} is nullable and not a boolean with a default, because the unique index
 * over it is what enforces "at most one active" — and a unique index counts NULLs as distinct.
 * Deactivation therefore writes NULL, never false.
 *
 * <p>{@code activationNote} records what the operator was shown when they confirmed, which is
 * what makes "why did four hundred issues close that afternoon" answerable later.
 */
@Entity
@Table(name = "t_semgrep_rule_set")
public class SemgrepRuleSetEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Long id;

    @Column(name = "name", length = 255, nullable = false)
    private String name;

    @JdbcTypeCode(SqlTypes.LONGVARCHAR)
    @Column(name = "files", nullable = false)
    private String files;

    @Column(name = "content_hash", length = 64, nullable = false)
    private String contentHash;

    @Column(name = "rule_count", nullable = false)
    private int ruleCount;

    @Column(name = "file_count", nullable = false)
    private int fileCount;

    @Column(name = "size_bytes", nullable = false)
    private Long sizeBytes;

    @Column(name = "is_active")
    private Boolean isActive;

    @Column(name = "uploaded_by", length = 255)
    private String uploadedBy;

    @Column(name = "uploaded_at", nullable = false)
    private Instant uploadedAt;

    @JdbcTypeCode(SqlTypes.LONGVARCHAR)
    @Column(name = "activation_note")
    private String activationNote;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getFiles() {
        return files;
    }

    public void setFiles(String files) {
        this.files = files;
    }

    public String getContentHash() {
        return contentHash;
    }

    public void setContentHash(String contentHash) {
        this.contentHash = contentHash;
    }

    public int getRuleCount() {
        return ruleCount;
    }

    public void setRuleCount(int ruleCount) {
        this.ruleCount = ruleCount;
    }

    public int getFileCount() {
        return fileCount;
    }

    public void setFileCount(int fileCount) {
        this.fileCount = fileCount;
    }

    public Long getSizeBytes() {
        return sizeBytes;
    }

    public void setSizeBytes(Long sizeBytes) {
        this.sizeBytes = sizeBytes;
    }

    public Boolean getIsActive() {
        return isActive;
    }

    public void setIsActive(Boolean isActive) {
        this.isActive = isActive;
    }

    public String getUploadedBy() {
        return uploadedBy;
    }

    public void setUploadedBy(String uploadedBy) {
        this.uploadedBy = uploadedBy;
    }

    public Instant getUploadedAt() {
        return uploadedAt;
    }

    public void setUploadedAt(Instant uploadedAt) {
        this.uploadedAt = uploadedAt;
    }

    public String getActivationNote() {
        return activationNote;
    }

    public void setActivationNote(String activationNote) {
        this.activationNote = activationNote;
    }
}
